Add keep3r guide (#10)
OpenZeppelin/defender-docs: docs/modules/ROOT/pages/guide-keep3r.adoc

> # Setting up a keep3r

Reviewer: ```suggestion
# Keep3r Network setup
```

> This guide will walk you through the process of running a Keeper in the https://keep3r.network/[keep3r.network] by leveraging Defender xref:relay.adoc[Relays] and xref:autotasks.adoc[Autotasks].

Reviewer: ```suggestion
This guide will walk you through the process of running a Keeper in the https://keep3r.network/[Keep3r Network] by leveraging Defender xref:relay.adoc[Relay] and xref:autotasks.adoc[Autotasks].
```

WARNING: The keep3r.network project is in beta. Use at your own risk.

[[what-is-a-keeper]]
## What is a Keeper?

> In the context of the keep3r network, Keepers are participants who execute available jobs in exchange for a reward. A contracts may register a job to be executed, and once it's available, a registered Keeper can send a transaction to execute it.

Reviewer: ```suggestion
In the context of the Keep3r Network, Keepers are participants who execute available jobs in exchange for a reward. A contract may register a job to be executed, and once it's available, a registered Keeper can send a transaction to execute it.
```

[quote, 'https://docs.keep3r.network/#keepers[keep3r.network documentation]']
____

> A Keeper is the term used to refer to an external person and/or team that executes a job. This can be as simplistic as calling a transaction, or as complex as requiring extensive off-chain logic. The scope of Keep3r network is not to manage these jobs themselves, but to allow contracts to register as jobs for keepers, and keepers to register themselves as available to perform jobs. It is up to the individual keeper to set up their devops and infrastructure and create their own rules based on what transactions they deem profitable.

Reviewer: This is not formatted as a quote.  

@frangio how do we format this as a quote?

Reviewer 2: The syntax is correct. The docs site is not styling it as a quote. Will take a look at that.

Reviewer 2: Should be better the next time the deploy preview is rendered.

____

[[prerequsites]]
## Prerequisites

To get started, you'll need the following:

> . *A Defender mainnet account*. Head over to https://defender.openzeppelin.com/[Defender] to sign up for a new account, and then https://openzeppelin.com/apply/[apply for mainnet access] indicating that you are interested in running a keep3r. We'll contact you shortly to activate your account for mainnet access. Note that mainnet accounts are free for now, but will be paid starting end of year.

Reviewer: ```suggestion
. *A Defender mainnet account*. Head over to https://defender.openzeppelin.com/[Defender] to sign up for a new account, and then https://openzeppelin.com/apply/[apply for mainnet access] indicating that you are interested in running a Keeper. We'll contact you shortly to activate your account for mainnet access. Note that mainnet accounts are free for now, but will be paid starting end of year.
```

> . *A Defender mainnet account*. Head over to https://defender.openzeppelin.com/[Defender] to sign up for a new account, and then https://openzeppelin.com/apply/[apply for mainnet access] indicating that you are interested in running a keep3r. We'll contact you shortly to activate your account for mainnet access. Note that mainnet accounts are free for now, but will be paid starting end of year.

Reviewer: Do we have a specific time frame for when accounts will be chargeable?

Author: December/January, still TBD

. *ETH to pay for transactions gas*. Keeper transactions require gas for execution, so make sure to have some ETH available to fund your Relay.

> . *An Alchemy, Etherscan, and/or Infura key (optional)*. Running a Keeper requires querying which jobs are available, and https://dashboard.alchemyapi.io/signup?referral=53fcee38-b894-4d5f-bd65-885d241f8d29[Alchemy], https://infura.io/[Infura], and https://etherscan.io/apis[Etherscan] provide free access to the network for executing these queries. You can do without them, but you may get throttled during your Keepers execution. Read more about this https://docs.ethers.io/v5/api-keys/[here].

Reviewer: Explicitly refer to referral link (we do this in the connect to public testnets Learn guide)

```suggestion
. *An Alchemy, Etherscan, and/or Infura key (optional)*. Running a Keeper requires querying which jobs are available, and https://dashboard.alchemyapi.io/signup?referral=53fcee38-b894-4d5f-bd65-885d241f8d29[Alchemy] (includes referral code), https://infura.io/[Infura], and https://etherscan.io/apis[Etherscan] provide free and paid access to the network for executing these queries. You can do without an API key, but you may get throttled during your Keepers execution. Read more about this https://docs.ethers.io/v5/api-keys/[here].
```

## Creating a Relayer

The first step will be to create a Defender Relayer to act as your Keeper. This Relayer will be the Ethereum account you will be executing jobs from. 

> To create a Relayer, simple head over to the Relayer section in Defender, and create a new one on the Mainnet network.

Reviewer: ```suggestion
To create a Relayer, simple head over to the xref:relay.adoc[Relay] section in Defender, and create a new relayer on the Mainnet network.
```

WARNING: If the option to create the relayer in mainnet is locked, it means your Defender account does not have mainnet access yet. Make sure to https://openzeppelin.com/apply/[apply for mainnet access first] and wait for our response.

## Fund your Relayer

> Next step is to send ETH to your Relayer, so it has funds to execute the jobs. Copy the Ethereum address of your new Relayer, and use your favorite wallet to send ETH to it. You can send whatever amount you want, but you want at least around 0.2 ETH to be able to send a handful of transactions before running out.

Author: Agos, Martin, and Igor are working on a UI that will make it easy - same as withdrawing tokens.

NOTE: Defender will send you an email when your Relay's funds drop below 0.1 ETH to remind you to refill it.

## Create an Autotask

Now that you have your Relay ready to send transactions, let's set up the Defender Autotask that will instruct the Relay to send them. An Autotask is a snippet of code that runs at regular intervals, in this case, we'll use it to automate the process of bonding, activating, searching for jobs, and executing them.

> Create a new Autotask in the Autotasks section of Defender, choosing to run every few minutes, and connecting it to the Relay you had created previously.

Reviewer: ```suggestion
Create a new Autotask in the xref:autotasks.adoc[Autotasks] section of Defender, choosing to run every few minutes, and connecting it to the Relay you had created previously.
```

> Create a new Autotask in the Autotasks section of Defender, choosing to run every few minutes, and connecting it to the Relay you had created previously.

Reviewer: Can we  be more specific than every few minutes?  What frequency can they use within their plan?

> 

Reviewer: We should add an explanation that the code below registers a Keeper and once activated will attempt to do work for the `YearnV1EarnKeep3r` as an example.

> As for the code, paste the following snippet, replacing the `0xAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA` value with your Relay address in the first line, and replacing `NONE` with the Alchemy, Infura, or Etherscan tokens respectively if you have any.

Reviewer: Should we be more specific?  Tokens may be a confusing term (may get confused with ERC20 tokens)

IIRC:
Alchemy is an API key
Infura is a Project ID
Etherscan is an API key

Reviewer 3: Agreed here. Tokens isn't precise enough for new comer and doesn't refer to what a user see. For example, with Alchemy, "Token" or "API Key" doesn't exist... after we have created a free account, we have only two choices (i) "HTTP URL" or (ii) "WEBSOCKETS URL". Using the exact terms given by the services (or put an example with the same format, that we will have to change as 0xAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA )  is less confusing for new comer I think.

Author: Good points!

[source,jsx]
----
// Replace with the Relay address

> const from = '0xAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA';

Reviewer: Could we add comments around the section that Keepers could change, the from address and the tokens, and a comment around the section that they don't need to change?

Author: Will do! But it's possible that this is no longer necessary :-)

  const registry = new ethers.Contract(registryAddr, registryAbi, signer);
  const contract = new ethers.Contract(workableAddr, workableAbi, signer);

  // Work if is a registered keeper
  if (await registry.isKeeper(from)) {
    return await workIfNeeded(contract);
  }

  // Otherwise run through bonding and activation process
  const bonding = await registry.bondings(from, collateralAddr).then(b => b.toNumber());
  if (bonding === 0) {
    const tx = await registry.bond(collateralAddr, collateralAmount);
    console.log(`Bonded relayer: ${tx.hash}`);
  } else if (bonding < Date.now() / 1000) {
    const tx = await registry.activate(collateralAddr);
    console.log(`Activated relayer: ${tx.hash}`);
    await workIfNeeded(contract);
  } else {
    console.log(`Waiting ${bonding - parseInt(Date.now() / 1000)} seconds until activation is available`);
  }
}

// To run locally (this code will not be executed in Autotasks)
if (require.main === module) {
  const { API_KEY: apiKey, API_SECRET: apiSecret } = process.env;
  exports.handler({ apiKey, apiSecret })
    .then(() => process.exit(0))
    .catch(error => { console.error(error); process.exit(1); });
}
----

> ## Wait for it!

Reviewer: ```suggestion
## Wait for Keeper activation
```

Author: Boooo https://www.youtube.com/watch?v=O_mlJwQ1_ZM

> The script above will take care of *bonding* your Relay to the keep3r network using zero KPR tokens as collateral. After bonding, there is a three-day delay enforced by the network before you can start using your Keeper. Once that time has passed, the script above will take care of *activating* your Relay as a Keeper.

Reviewer: Is this required per Keeper task or is it a single registration with the Keeper Network?

If it is a single registration, could we split the registration into its own Autotask?  
If it is a registration per Keeper task, then we should add a note of this.  We could also put comments around the registration code so that this can be used again for other tasks.

Author: We are working to handle registration directly on the Relayer, so this is does not pollute the Autotask. For now, I preferred to include everything in a single autotask to simplify setup - user needs to copy-paste only one code snippet.

Author: That said, agree on the comments.

Once your Keeper is activated, it will query the `YearnV1EarnKeep3r` job to see if there is work to do by calling `workable`, and if so, will send a transaction to execute the `work`. You can check out your Autotask logs to see what your Keeper has done on each run.

> WARNING: Keep in mind that sending the transaction does not automatically mean you will get rewarded for it: other keepers may also be monitoring the job for work, and if their transaction gets mined before yours, you will not get the reward!

Reviewer: ```suggestion
WARNING: Keep in mind that sending the transaction does not automatically mean you will get rewarded for it: other Keepers may also be monitoring the job for work, and if their transaction gets mined before yours, you will not get the reward!
```

## Coming soon...

> We are working on additional features in Defender to make this process easier. This involves being able to send Relayer transactions from the Defender site, so you can bond the Relay with the collateral you want and easily transfer your KPR rewards out of the Relay to your wallet without having to write code for it; adding a custom Defender provider, so you don't need to request an Alchemy, Infura, or Etherscan token yourself; creating a registry of Autotask scripts, so you don't need to copy-paste code from this document; showing Keeper information in the Relayer page, so you know whether your Relayer has been activated and the rewards earned so far; and more. We will be updating this document as we progress with these features. Stay tuned!

Reviewer: We should make this a bullet point list
